examples/petsc/bp3.h

```diff
@@ -76,13 +78,29 @@ CEED_QFUNCTION(Diff)(void *ctx, const CeedInt Q,
   const CeedScalar *ug = in[0], *qd = in[1];
   CeedScalar *vg = out[0];
 
+  // Quadrature Point Loop
+  CeedPragmaOMP(simd)
   for (CeedInt i=0; i<Q; i++) {
-    const CeedScalar ug0 = ug[i+Q*0];
-    const CeedScalar ug1 = ug[i+Q*1];
-    const CeedScalar ug2 = ug[i+Q*2];
-    vg[i+Q*0] = qd[i+Q*0]*ug0 + qd[i+Q*1]*ug1 + qd[i+Q*2]*ug2;
-    vg[i+Q*1] = qd[i+Q*1]*ug0 + qd[i+Q*3]*ug1 + qd[i+Q*4]*ug2;
-    vg[i+Q*2] = qd[i+Q*2]*ug0 + qd[i+Q*4]*ug1 + qd[i+Q*5]*ug2;
-  }
+    // Read spatial derivatives of u components
+    const CeedScalar uJ[3]        =  {ug[i+Q*0],
+                                      ug[i+Q*1],
+                                      ug[i+Q*2]};
+    // Read qdata
+    const CeedScalar dXdxdXdxT[6] =  {qd[i+0*Q],
+                                      qd[i+1*Q],
+                                      qd[i+2*Q],
+                                      qd[i+3*Q],
+                                      qd[i+4*Q],
+                                      qd[i+5*Q]
+                                     };
+
+    const CeedInt idx[3][3] = {{0, 1, 2}, {1, 3, 4}, {2, 4, 5}}; // symmetric matrix indices
+
+    for (int j=0; j<3; j++) {
+      vg[i+(0+j*3)*Q]  = 0;
+      for (int k=0; k<3; k++)
+        vg[i+(0+j*3)*Q] += uJ[k] * dXdxdXdxT[idx[j][k]];
```

Comment:
Since the PETSc BPs are our benchmarks, I think it would be good to verify that this commit doesn't change our performance. I know it shouldn't, but it seems worth the peace of mind to double check.

Reply:
I have compared the benchmarks. Here you can find the bp3 currently on Master:
https://bit.ly/2kdVguU
and the new version with the loop:
https://bit.ly/2kG7RqX
Same comparison for bp4, current Master version:
https://bit.ly/2kG7VXJ
And version with the loop:
https://bit.ly/2lN1l1W

Reply:
For small numbers of local elements, it seems a bit faster with the loop. For larger number of elements there seems to be only a bit of noise between the two versions

Reply:
You can also see the plots for BP3 Master version:
http://bit.ly/Masterplot
Loop version:
http://bit.ly/LoopVersion

Same for BP4 Master version:
http://bit.ly/MasterBP4
Loop Version:
http://bit.ly/LoopVersionBP4